presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java

```diff
@@ -13,70 +13,181 @@
+        MethodHandle valueProcessor = getProcessMethodHandle(functionRegistry, fromValueType, toValueType, false);
+        MethodHandle target = MethodHandles.insertArguments(METHOD_HANDLE, 0, keyProcessor, valueProcessor, toMapType);
+        return new ScalarFunctionImplementation(true, ImmutableList.of(valueTypeArgumentProperty(RETURN_NULL_ON_NULL)), target, true);
+    }
+
+    private MethodHandle getProcessMethodHandle(FunctionRegistry functionRegistry, Type fromType, Type toType, boolean isKey)
+    {
+        MethodHandle getter = nativeValueGetter(fromType);
+
+        // Adapt cast that takes ([ConnectorSession,] ?) to one that takes (?, ConnectorSession), where ? is the return type of getter.
+        MethodHandle cast = functionRegistry.getScalarFunctionImplementation(internalOperator(CAST, toType, ImmutableList.of(fromType))).getMethodHandle();
+        if (cast.type().parameterArray()[0] != ConnectorSession.class) {
+            cast = MethodHandles.dropArguments(cast, 0, ConnectorSession.class);
+        }
+        cast = permuteArguments(cast, methodType(cast.type().returnType(), cast.type().parameterArray()[1], cast.type().parameterArray()[0]), 1, 0);
+        MethodHandle target = compose(cast, getter);
+
+        // We only need to check null if the method handle is for key and the cast result is nullable.
+        Class<?> castResultJavaType = target.type().returnType();
+        if (isKey && !castResultJavaType.isPrimitive()) {
+            MethodHandle nullCheck;
+            if (castResultJavaType == Long.class) {
+                nullCheck = methodHandle(MapToMapCast.class, "checkLongIsNotNull", Long.class);
+            }
+            else if (castResultJavaType == Double.class) {
+                nullCheck = methodHandle(MapToMapCast.class, "checkDoubleIsNotNull", Double.class);
+            }
+            else if (castResultJavaType == Boolean.class) {
+                nullCheck = methodHandle(MapToMapCast.class, "checkBooleanIsNotNull", Boolean.class);
+            }
+            else if (castResultJavaType == Slice.class) {
+                nullCheck = methodHandle(MapToMapCast.class, "checkSliceIsNotNull", Slice.class);
+            }
+            else if (castResultJavaType == Block.class) {
+                nullCheck = methodHandle(MapToMapCast.class, "checkBlockIsNotNull", Block.class);
+            }
+            else {
+                throw new IllegalArgumentException("Unknown java type " + castResultJavaType + " from type " + toType);
+            }
+            target = compose(nullCheck, target);
+        }
+
+        MethodHandle writer = nativeValueWriter(toType);
+        writer = permuteArguments(writer, methodType(void.class, writer.type().parameterArray()[1], writer.type().parameterArray()[0]), 1, 0);
+
+        // The composed method handle could be invoked by: (fromMap, index, connectorSession, blockBuilder)
```

Comment:
Put this in the javadoc of the function instead.

In addition, you want to expand it a bit. For example, `The signature of the returned MethodHandle is (Block fromMap, int position, ConnectorSession session, BlockBuilder mapBlockBuilder)void`